Sources/NIOExtras/HTTP1ProxyConnectHandler.swift

```diff
@@ -0,0 +1,399 @@
+        }
+
+        /// Proxy response contains unexpected body
+        public static func invalidProxyResponse(file: String = #file, line: UInt = #line) -> Error {
+            Error(error: .invalidProxyResponse, file: file, line: line)
+        }
+
+        /// Connection has been closed for ongoing request
+        public static func remoteConnectionClosed(file: String = #file, line: UInt = #line) -> Error {
+            Error(error: .remoteConnectionClosed, file: file, line: line)
+        }
+
+        /// Proxy connection handshake has timed out
+        public static func httpProxyHandshakeTimeout(file: String = #file, line: UInt = #line) -> Error {
+            Error(error: .httpProxyHandshakeTimeout, file: file, line: line)
+        }
+
+        /// Handler was removed before we received a result for the request
+        public static func noResult(file: String = #file, line: UInt = #line) -> Error {
+            Error(error: .noResult, file: file, line: line)
+        }
+
+        fileprivate var errorCode: Int {
+            switch self.store.details {
+            case .proxyAuthenticationRequired:
+                return 0
+            case .invalidProxyResponseHead:
+                return 1
+            case .invalidProxyResponse:
+                return 2
+            case .remoteConnectionClosed:
+                return 3
+            case .httpProxyHandshakeTimeout:
+                return 4
+            case .noResult:
+                return 5
+            }
+        }
+    }
+
+}
+
+extension NIOHTTP1ProxyConnectHandler.Error: Hashable {
+    // compare only the kind of error, not the associated response head
+    public static func == (lhs: Self, rhs: Self) -> Bool {
+        switch (lhs.store.details, rhs.store.details) {
+        case (.proxyAuthenticationRequired, .proxyAuthenticationRequired):
+            return true
+        case (.invalidProxyResponseHead, .invalidProxyResponseHead):
+            return true
+        case (.invalidProxyResponse, .invalidProxyResponse):
+            return true
+        case (.remoteConnectionClosed, .remoteConnectionClosed):
+            return true
+        case (.httpProxyHandshakeTimeout, .httpProxyHandshakeTimeout):
+            return true
+        case (.noResult, .noResult):
+            return true
+        default:
+            return false
+        }
+    }
+
+    public func hash(into hasher: inout Hasher) {
+        hasher.combine(self.errorCode)
+    }
+}
+
+
+extension NIOHTTP1ProxyConnectHandler.Error: CustomStringConvertible {
+    public var description: String {
+        self.store.details.description
+    }
+}
+
+extension NIOHTTP1ProxyConnectHandler.Error.Details: CustomStringConvertible {
+    public var description: String {
+        switch self {
+        case .proxyAuthenticationRequired:
+            return "Proxy Authentication Required"
+        case .invalidProxyResponseHead:
+            return "Invalid Proxy Response Head"
```

Comment:
you want to include the `HTTPResponseHead` in the description e.g.: 
```suggestion
        case .invalidProxyResponseHead(let head):
            return "Invalid Proxy Response Head \(head)"
```